docs/configuration.md
Configuration files
===================

All settings of Lensed, from the input files to the physical model of the lens
system, can be controlled from a single configuration file. A typical layout is
the following:

```ini
image       = double-lens.fits
gain        = 2400
offset      = 2.9633

output      = true
root        = chains/double-lens-

ds9         = true

[objects]
lens        = sie
source      = sersic

[priors]
lens.x      = unif 48 52
lens.y      = unif 48 52
lens.r      = unif 15 20
lens.q      = unif 0 1
lens.pa     = wrap unif 0 180
source.x    = image unif 80 85
source.y    = image unif 35 40
source.r    = unif 1 5
source.mag  = unif -5 0
source.n    = unif 0.5 8.0
source.q    = unif 0 1
source.pa   = wrap unif 0 180

[labels]
lens.x      = x_L
lens.y      = y_L
lens.r      = r_L
lens.q      = q_L
lens.pa     = \theta_L
source.x    = x_S
source.y    = y_S
source.r    = r_S
source.mag  = mag_S
source.n    = n_S
source.q    = q_S
source.pa   = \theta_S
```

There are four sections in a typical configuration file: program options (top),
the definition of the objects in the lens system, priors for the individual
parameters of the model, and optionally a number of labels for output.


Options
-------

Options can be given in three places: as command line arguments, at the
top of the configuration file, or later in a group called `[options]`.

The following options are known to Lensed (see `lensed --help`):

Option     | Type           | Description                            | Default
-----------|----------------|----------------------------------------|--------
`device`   | `string`       | [Select computation device.](#device)  | `auto`
`output`   | `bool`         | Output results.                        | `true`
`root`     | `string`       | Root element for all output paths.     | 
`image`    | `path`         | Input image, FITS file in counts/sec.  | 
`gain`     | `real`, `path` | [Conversion factor to counts.](#gain)  | 
`offset`   | `real`         | Subtracted flat-field offset.          | `0`
`weight`   | `real`, `path` | Pixel weights in 1/(counts/sec)^2.     | `none`
`xweight`  | `real`, `path` | Extra weight map multiplier.           | `none`
`mask`     | `path`         | Input mask, FITS file.                 | `none`
`psf`      | `path`         | Point-spread function, FITS file.      | `none`
`rule`     | `string`       | Rule for numerical integration.        | `g3k7`
`nlive`    | `int`          | Number of live points.                 | `300`
`ins`      | `bool`         | Use importance nested sampling.        | `true`
`mmodal`   | `bool`         | Mode separation (if ins = false).      | `true`
`ceff`     | `bool`         | Constant efficiency mode.              | `true`
`acc`      | `real`         | Target acceptance rate.                | `0.05`
`tol`      | `real`         | Tolerance in log-evidence.             | `0.1`
`shf`      | `real`         | Shrinking factor.                      | `0.8`
`maxmodes` | `int`          | Maximum number of expected modes.      | `100`
`feedback` | `bool`         | Show feedback from MultiNest.          | `false`
`updint`   | `int`          | Update interval for output.            | `100`
`seed`     | `int`          | Random number seed for sampling.       | `-1`
`resume`   | `bool`         | Resume from last checkpoint.           | `false`
`maxiter`  | `int`          | Maximum number of iterations.          | `0`

Options without a default value are required to be set in the configuration.

If (and only if) XPA support is enabled in the [build options](building.md),
there are a number of additional settings for integration with SAOImage DS9:

Option     | Type           | Description                            | Default
-----------|----------------|----------------------------------------|--------
`ds9`      | `bool`         | Integrate with SAOImage DS9.           | `false`
`ds9-name` | `string`       | XPA template name for DS9.             | `ds9`

If region file support is enabled in the [build options](building.md), the
`mask` option can alternatively contain the path to a supported region file.

There are a number of caveats regarding the following options.

### device

The default option for `device` is `auto`. This option will select the first
GPU device, in case one exists. Alternatively, it will select the first device
found.

### gain

The effective gain can be given either as a real number, in which case it will
be applied uniformly for each pixel of the image, or the path to a FITS file
that contains the effective gain for each individual pixel. This can be, for
example, the `EXP` image extension of a file generated by MultiDrizzle.


Objects
-------

Objects are the individual components that create the physical model for the
reconstruction. Each object corresponds to a definition in the `kernel` folder
that lists its parameters and physical properties.

The model used for reconstruction is created by listing one or more objects in
the `[objects]` section of the configuration file, together with a unique name
for identification.

**Important:** The order of objects in the configuration file determines the
physical layout of the system. If a source is meant to be placed before/behind
a lens, it must appear in the list of objects above/below that lens.

Example:

```ini
[objects]
lens   = sis
source = sersic
```

This model contains a SIS lens called `lens` and a Sérsic source called
`source`. The source appears behind the lens and will thus be deflected by it.

```ini
[objects]
host   = sersic
lens   = sis
source = sersic
```

The same as above, but including a foreground Sérsic source hosted on the lens
plane. As the `host` object appears before the lens, it will not be deflected.


Priors
------

Priors for parameters are specified in section `[priors]` of a configuration
file. They are given in the format

```ini
[priors]
obj.param = [wrap] [image] <prior> <arg0> <arg1> ...
```

An exception is the pseudo-prior that fixes the value of a parameter, which is
given simply as `<value>`, without any name.

The optional `wrap` prefix can be used to indicate that the parameter values
wrap around at the boundaries of the prior. This is useful to get the correct
distribution for cyclic parameters such as the orientation, where a mean value
of 170 ± 20 degrees falls out of the natural [0, 180] degree bounds.

The following priors are known:

Prior          | Description
---------------|--------------------------------------------------------------
`<value>`      | pseudo-prior that fixes the parameter to the given *value*
`unif <a> <b>` | uniform prior on the interval [*a*, *b*]
`norm <m> <s>` | normal prior with mean *m* and standard deviation *s*

Additionally, the following optional keywords can be specified:

Keyword        | Description
---------------|--------------------------------------------------------------
`wrap`         | prior is cyclic and wraps around at the boundaries
`image`        | prior is specified on the image plane

The `image` keyword can be used to specify a prior on the observed quantities
on the image plane. This is limited to position parameters.

Example:

```ini
[priors]

; parameter "x" of object "lens" is fixed to 100
lens.x = 100

; normal prior for parameter "y" of object "lens" with mean 100 and sigma 20
lens.y = norm 100 20

; uniform probability for parameter "r" of object "lens" in interval [10, 20]
lens.r = unif 10 20

; the orientation angle is between 0 and 180 degrees and wraps around
lens.pa = wrap unif 0 180

; the source position is given by one of its images
source.x = image unif 80 85
source.y = image unif 35 40
```


Labels
------

It is possible to attach labels to parameters for post-processing e.g. with
getdist. These labels are given in the `[labels]` section of a configuration
file. The format is

```ini
[labels]
obj.param = <label>
```

Example:

```ini
[labels]

; label parameter "x" of object "lens"
lens.x = x_L
```


CHANGELOG.md
Changelog
=========

v1.3.1 (2017-03-15)
-------------------

  * fix OpenCL compile errors due to INFINITY macro

v1.3.0 (2017-03-14)
-------------------

  * fix OpenCL compatibility issues when building
  * new `bscale` option for config file
  * support for even PSF size
  * weight, xweight, gain options take file or number
  * Sersic approximations for full parameter range
  * fix crash when no gain option is given
  * output p-value map
  * use CFITSIO to check whether file is FITS
  * use region files for mask
  * default values of zero for sky gradient
  * fixed existing EPL tests and added isothermal test
  * write paramnames and ranges at start of reconstruction
  * 7-sigma bounds instead of INF for normal prior
  * simplified Sersic source coefficients

v1.2.0 (2016-02-12)
-------------------

  * FITS extension names and weight map output
  * check that weights are positive
  * boolean options can be set as flags
  * XPA support and basic DS9 integration

v1.1.1 (2016-01-12)
-------------------

  * bugfix: use parameter map to get correct wraparound flag
  * fixed bug when last object was a lens
  * fixed EPL scale

v1.1.0 (2015-10-10)
-------------------

  * parameters can have default values
  * pseudo-priors are derived parameters
  * quadrature rule selection
  * suppress file output from Fortran in quiet or batch mode
  * image plane priors for position
  * object data in local memory
  * show maxloglike in progress
  * support for extra weight map
  * elliptical power law profile lens
  * show progress in MultiNest feedback
  * use cl_float instead of cl_char for object memory

v1.0.2 (2015-06-26)
-------------------

  * catch keyboard interrupts and print results
  * minor output tweak for help display and device list
  * simple profiling for OpenCL functions
  * parameter bounds
  * parameter types
  * add Makefile option for debug builds
  * parameter wrap specification with prior
  * added parameter file and documentations for analysis with GetDist

v1.0.1 (2015-06-05)
-------------------

  * improved handling of binary dependencies
  * documentation for releases
  * show feedback from MultiNest

v1.0.0 (2015-05-28)
-------------------

  * changed spelling of Lensed
  * documentation for Lensed
  * platform support
  * smarter object definitions
  * MIT license added
  * fix invalid memory being freed by options
  * fix infinite deflections
  * fixed bug when no default config was present
  * compatibility with older CFITSIO
  * added OpenCL to list of requirements
  * support for default config file
  * OpenCL device selection
  * New example with simultaneous optimization of lens mass, lens light and
    background source
  * Added extras/pl_density2d.pro for simple visualisation using IDL.
  * tests for existing objects
  * Inverting the directions of the psf convolution
  * fix FWHM estimation of background counts
  * normal distribution prior
  * Individual work sizes for kernels.
  * Fix the blank log files for MultiNest.
  * Limit block size by OpenCL device capabilities.
  * Flag to list computation devices.
  * Pixel coordinate system.
  * Added Elliptical Power-Law Potential models, eplp.cl and eplp_plus_shear.cl
  * Optimised and unified object code.
  * Added sis_plus_shear and sie_plus_shear lens models. Removed shear model.
  * Define kernel constants also in object code.
  * Added some helpful links and PR info.
  * Fix line-too-long if there's no final newline in ini.
  * Fixes incorrect allocation of PSF buffer.
  * Added point mass lens.
  * Split `efr` option into separate `acc` and `shf` options.
  * Functions for and output of prior bounds.
  * New lens models: sis, nsis, nsie
  * Adding new source models.
  * Fix ambiguous `mad24` overload in kernel. Clean up Makefile.

v0.4.0 (2015-01-12)
-------------------

  * Replace PCC by GK integration rule.
  * Parameterisation of models.
  * Read per-pixel effective gain from FITS file.
  * Make sure no second lensing plane is created.

v0.3.0 (2014-12-10)
-------------------

  * Output of residuals and relative errors.
  * Convolution with PSF.
  * Separate kernels for rendering and likelihood.
  * Automatic freeing of options.
  * Paths are read relative to input files.
  * Better log file handling.
  * Warn if background is zero but no offset was given.
  * Warn user if a sky should be used.
  * Foreground objects, sky.
  * Delta function prior.
  * Disentangled options and parser.
  * Output numerical constants.
  * Simplified data handling with optional weight map.
  * Linear kernel execution.
  * Output kernel device information.
  * Output kernel notifications to stderr.
  * Kernel works in two dimensions.
  * Added info on how to contribute.
  * Example added.

v0.2.0 (2014-11-30)
-------------------

  * Some fixes for Makefile when releasing versions.
  * Release tool for updating version number.
  * Don't wait for user interaction in batch mode. Closes #51.
  * Fixed memory leak when redefining a prior.
  * Batch output mode.
  * Parameters have `id` field in the form "x.y".
  * Moved internal `errno`-related errors to `errori` function.
  * Write a getdist-compatible `<root>.paramnames` file.
  * Dumper outputs images and saves results.
  * Organised parameter space creation.
  * Options are required depending on other options.
  * Correct sign for Sérsic magnitude.
  * Parse numerical constants such as `pi`.
  * Various fixes, mostly memory leaks.
  * Priors added.
  * Input parsing factored into its own module.
  * GPU option added.
  * Minor output fixes.
  * More simplifications of kernel code for objects.
  * Output generated kernel code.
  * Simpler check for duplicate objects in kernel creation.
  * Dynamic size of object data.
  * Dynamically generate kernel code to set parameters.
  * Dynamically generate kernel code to compute surface brightness.
  * Get object meta-data from kernels.
  * Simplified object definition in kernel code.

src/input/options.c
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "../input.h"
#include "options.h"
#include "../parse.h"
#include "../log.h"

// information about input option
struct option
{
    const char* name;
    const char* help;
    int required;
    size_t depend;
    const char* type;
    int (*read)(void*, const char*);
    int (*write)(char*, const void*, size_t);
    void (*free)(void*);
    union {
        void* default_null;
        const char* default_string;
        int default_bool;
        int default_int;
        double default_real;
        const char* default_path;
        struct path_or_real* default_path_or_real;
    } default_value;
    size_t offset;
    size_t size;
};

// mark option as required or optional
#define OPTION_REQUIRED(type) 1, 0, #type, option_read_##type, option_write_##type, option_free_##type, { .default_null = NULL }
#define OPTION_OPTIONAL(type, value) 0, 0, #type, option_read_##type, option_write_##type, option_free_##type, { .default_##type = value }
#define OPTION_REQIFSET(type, value, depend) 2, offsetof(options, depend), #type, option_read_##type, option_write_##type, option_free_##type, { .default_##type = value }
#define OPTION_REQIFNOT(type, value, depend) 3, offsetof(options, depend), #type, option_read_##type, option_write_##type, option_free_##type, { .default_##type = value }

// get offset of option field in input
#define OPTION_FIELD(field) offsetof(options, field), sizeof(((options*)0)->field)

// get status of dependency
#define OPTION_ISSET(options, offset) (*(int*)((char*)options + offset))
#define OPTION_ISNOT(options, offset) (!*(int*)((char*)options + offset))

// declare option types
#define OPTION_TYPE(type) \
    static int option_read_##type(void*, const char*); \
    static int option_write_##type(char*, const void*, size_t); \
    static void option_free_##type(void*);

// list known option types
OPTION_TYPE(string)
OPTION_TYPE(bool)
OPTION_TYPE(int)
OPTION_TYPE(real)
OPTION_TYPE(path)
OPTION_TYPE(path_or_real)

// list of known options
struct option OPTIONS[] = {
    {
        "device",
        "Select computation device",
        OPTION_OPTIONAL(string, "auto"),
        OPTION_FIELD(device)
    },
    {
        "output",
        "Output results",
        OPTION_OPTIONAL(bool, 1),
        OPTION_FIELD(output)
    },
    {
        "root",
        "Root element for all output paths",
        OPTION_REQIFSET(string, NULL, output),
        OPTION_FIELD(root)
    },
    {
        "image",
        "Input image, FITS file in counts/sec",
        OPTION_REQUIRED(path),
        OPTION_FIELD(image)
    },
    {
        "gain",
        "Conversion factor to counts",
        OPTION_REQIFNOT(path_or_real, NULL, weight),
        OPTION_FIELD(gain)
    },
    {
        "offset",
        "Subtracted flat-field offset",
        OPTION_OPTIONAL(real, 0),
        OPTION_FIELD(offset)
    },
    {
        "bscale",
        "Scale input pixel values",
        OPTION_OPTIONAL(real, 0),
        OPTION_FIELD(bscale)
    },
    {
        "weight",
        "Weight map in 1/(counts/sec)^2",
        OPTION_OPTIONAL(path_or_real, NULL),
        OPTION_FIELD(weight)
    },
    {
        "xweight",
        "Extra weight map multiplier",
        OPTION_OPTIONAL(path_or_real, NULL),
        OPTION_FIELD(xweight)
    },
    {
        "mask",
        "Input mask, FITS file",
        OPTION_OPTIONAL(path, NULL),
        OPTION_FIELD(mask)
    },
    {
        "psf",
        "Point-spread function, FITS file",
        OPTION_OPTIONAL(path, NULL),
        OPTION_FIELD(psf)
    },
    {
        "rule",
        "Rule for numerical integration",
        OPTION_OPTIONAL(string, "g3k7"),
        OPTION_FIELD(rule)
    },
#ifdef LENSED_XPA
    {
        "ds9",
        "Integrate with SAOImage DS9",
        OPTION_OPTIONAL(bool, 0),
        OPTION_FIELD(ds9)
    },
    {
        "ds9-name",
        "XPA template name for DS9",
        OPTION_OPTIONAL(string, "ds9"),
        OPTION_FIELD(ds9_name)
    },
#endif
    {
        "nlive",
        "Number of live points",
        OPTION_OPTIONAL(int, 300),
        OPTION_FIELD(nlive)
    },
    {
        "ins",
        "Use importance nested sampling",
        OPTION_OPTIONAL(bool, 1),
        OPTION_FIELD(ins)
    },
    {
        "mmodal",
        "Mode separation (if ins = false)",
        OPTION_OPTIONAL(bool, 1),
        OPTION_FIELD(mmodal)
    },
    {
        "ceff",
        "Constant efficiency mode",
        OPTION_OPTIONAL(bool, 1),
        OPTION_FIELD(ceff)
    },
    {
        "acc",
        "Target acceptance rate",
        OPTION_OPTIONAL(real, 0.05),
        OPTION_FIELD(acc)
    },
    {
        "tol",
        "Tolerance in log-evidence",
        OPTION_OPTIONAL(real, 0.1),
        OPTION_FIELD(tol)
    },
    {
        "shf",
        "Shrinking factor",
        OPTION_OPTIONAL(real, 0.8),
        OPTION_FIELD(shf)
    },
    {
        "maxmodes",
        "Maximum number of expected modes",
        OPTION_OPTIONAL(int, 100),
        OPTION_FIELD(maxmodes)
    },
    {
        "feedback",
        "Show feedback from MultiNest",
        OPTION_OPTIONAL(bool, 0),
        OPTION_FIELD(feedback)
    },
    {
        "updint",
        "Update interval for output",
        OPTION_OPTIONAL(int, 100),
        OPTION_FIELD(updint)
    },
    {
        "seed",
        "Random number seed for sampling",
        OPTION_OPTIONAL(int, -1),
        OPTION_FIELD(seed)
    },
    {
        "resume",
        "Resume from last checkpoint",
        OPTION_OPTIONAL(bool, 0),
        OPTION_FIELD(resume),
    },
    {
        "maxiter",
        "Maximum number of iterations",
        OPTION_OPTIONAL(int, 0),
        OPTION_FIELD(maxiter)
    }
};

// number of known options
#define NOPTIONS (sizeof(OPTIONS)/sizeof(struct option))

// last error message
static char ERROR_MSG[1024] = {0};

// current path for option reading
static const char* OPTIONS_CWD = NULL;

options* create_options()
{
    options* opts = malloc(sizeof(options));
    if(!opts)
        errori(NULL);
    memset(opts, 0, sizeof(options));
    return opts;
}

void free_options(options* opts)
{
    // TODO: free options only when they were set, and when they are reset
    //for(int i = 0; i < NOPTIONS; ++i)
    //    OPTIONS[i].free((char*)opts + OPTIONS[i].offset);
    free(opts);
}

void default_options(input* inp)
{
    for(int i = 0; i < NOPTIONS; ++i)
    {
        // copy default value to input
        memcpy((char*)inp->opts + OPTIONS[i].offset, &OPTIONS[i].default_value, OPTIONS[i].size);
        
        // set required flag for option
        inp->reqs[i] = OPTIONS[i].required;
    }
}

int read_option(input* inp, const char* name, const char* value)
{
    // read all chars in name
    return read_option_n(inp, name, strlen(name), value);
}

int read_option_n(input* inp, const char* name, int n, const char* value)
{
    int opt;
    
    // error if no name was given
    if(n == 0)
    {
        snprintf(ERROR_MSG, sizeof(ERROR_MSG)-1, "no option given");
        return 1;
    }
    
    // find option
    for(opt = 0; opt < NOPTIONS; ++opt)
        if(strncmp(name, OPTIONS[opt].name, n) == 0)
            break;
    
    // error if option was not found
    if(opt == NOPTIONS)
    {
        snprintf(ERROR_MSG, sizeof(ERROR_MSG)-1, "invalid option: %.*s", n, name);
        return 1;
    }
    
    // check if no value was given
    if(!value || !*value)
    {
        // booleans do not need value
        if(OPTIONS[opt].read == option_read_bool)
        {
            // enable boolean flag
            value = "true";
        }
        else
        {
            // error, no value given
            snprintf(ERROR_MSG, sizeof(ERROR_MSG)-1, "option %.*s: no value given", n, name);
            return 1;
        }
    }
    
    // try to read option and return eventual errors
    if(OPTIONS[opt].read((char*)inp->opts + OPTIONS[opt].offset, value))
    {
        snprintf(ERROR_MSG, sizeof(ERROR_MSG)-1, "option %.*s: invalid value: %s (should be %s)", n, name, value, OPTIONS[opt].type);
        return 1;
    }
    
    // option is no longer required
    inp->reqs[opt] = 0;
    
    // report success
    return 0;
}

const char* options_error()
{
    return ERROR_MSG;
}

size_t noptions()
{
    return NOPTIONS;
}

const char* option_name(size_t n)
{
    return OPTIONS[n].name;
}

const char* option_type(size_t n)
{
    return OPTIONS[n].type;
}

const char* option_help(size_t n)
{
    return OPTIONS[n].help;
}

int option_required(size_t n)
{
    return OPTIONS[n].required;
}

int option_resolved(size_t n, options* opts, int reqs[])
{
    int res = 0;
    
    if(reqs[n] == 0)
        res = 1;
    else if(reqs[n] == 1)
        res = 0;
    else if(reqs[n] == 2)
        res = !OPTION_ISSET(opts, OPTIONS[n].depend);
    else if(reqs[n] == 3)
        res = !OPTION_ISNOT(opts, OPTIONS[n].depend);
    
    return res;
}

int option_default_value(char* buf, size_t buf_size, size_t n)
{
    return OPTIONS[n].write(buf, &OPTIONS[n].default_value, buf_size);
}

int option_value(char* buf, size_t buf_size, const input* inp, size_t n)
{
    return OPTIONS[n].write(buf, (char*)inp->opts + OPTIONS[n].offset, buf_size);
}

const char* options_cwd(const char* cwd)
{
    const char* old = OPTIONS_CWD;
    OPTIONS_CWD = cwd;
    return old;
}

// wrappers for option reading and writing

int option_read_string(void* out, const char* in)
{
    char** out_char = out;
    *out_char = malloc(strlen(in) + 1);
    strcpy(*out_char, in);
    return 0;
}

int option_write_string(char* out, const void* in, size_t n)
{
    char* const* in_char = in;
    snprintf(out, n, "%s", *in_char ? *in_char : "none");
    return 0;
}

void option_free_string(void* p)
{
    char** str = p;
    free(*str);
}

int option_read_bool(void* out, const char* in)
{
    int* out_bool = out;
    return read_bool(out_bool, in);
}

int option_write_bool(char* out, const void* in, size_t n)
{
    const int* in_bool = in;
    return write_bool(out, *in_bool, n);
}

void option_free_bool(void* p)
{
}

int option_read_int(void* out, const char* in)
{
    int* out_int = out;
    return read_int(out_int, in);
}

int option_write_int(char* out, const void* in, size_t n)
{
    const int* in_int = in;
    return write_int(out, *in_int, n);
}

void option_free_int(void* p)
{
}

int option_read_real(void* out, const char* in)
{
    double* out_real = out;
    return read_real(out_real, in);
}

int option_write_real(char* out, const void* in, size_t n)
{
    const double* in_real = in;
    return write_real(out, *in_real, n);
}

void option_free_real(void* p)
{
}

int option_read_path(void* out, const char* in)
{
    int r;
    char* str;
    char* abs;
    
    // read string
    r = option_read_string(out, in);
    if(r != 0)
        return r;
    str = *(char**)out;
    
    // check for absolute path
    if(str[0] == '/')
        return 0;
    
    // check if there is a cwd for options
    if(OPTIONS_CWD == NULL)
        return 0;
    
    // append relative path to current path
    abs = malloc(strlen(OPTIONS_CWD) + strlen("/") + strlen(str) + 1);
    if(!abs)
        errori(NULL);
    sprintf(abs, "%s%s%s", OPTIONS_CWD, "/", str);
    
    // swap strings
    *(char**)out = abs;
    free(str);
    
    return 0;
}

int option_write_path(char* out, const void* in, size_t n)
{
    return option_write_string(out, in, n);
}

void option_free_path(void* p)
{
    char** str = p;
    free(*str);
}

int option_read_path_or_real(void* out, const char* in)
{
    int err;
    struct path_or_real** por = out;
    
    // allocate space for struct
    *por = malloc(sizeof(struct path_or_real));
    if(!*por)
        errori(NULL);
    
    // try to read real
    err = option_read_real(&(*por)->value, in);
    
    // check if there was a number
    if(!err)
    {
        // no file was given
        (*por)->file = NULL;
        return 0;
    }
    
    // no number was given
    (*por)->value = 0;
    
    // read path
    return option_read_path(&(*por)->file, in);
}

int option_write_path_or_real(char* out, const void* in, size_t n)
{
    struct path_or_real* const* por = in;
    if(!*por)
    {
        snprintf(out, n, "%s", "none");
        return 0;
    }
    if((*por)->file)
        return option_write_string(out, &(*por)->file, n);
    else
        return option_write_real(out, &(*por)->value, n);
}

void option_free_path_or_real(void* p)
{
    struct path_or_real** por = p;
    if(*por)
    {
        if((*por)->file)
            option_free_path(&(*por)->file);
        else
            option_free_real(&(*por)->value);
        free(*por);
    }
}
